File: data_structure/set/src/SetDemo.py
# Python
# データ構造: セット (Set)

class SetData:
    def __init__(self):
        self._data = []

    def get(self):
        # 要素を取得
        return self._data

    def get_index(self, item):
        # 指定された要素がセット内に存在するかどうかをチェックする。
        try:
            index = self._data.index(item)
            return index
        except ValueError:
            print(f"ERROR: {item} は範囲外です")
            return -1

    def get_value(self, index):
        # 指定されたインデックスの要素を取得する。
        if 0 <= index < len(self._data):
            return self._data[index]
        else:
            print(f"ERROR: {index} は範囲外です")
            return None

    def add(self, item):
        # 要素をセットに追加する。
        if item not in self._data:
            self._data.append(item)
            return True
        else:
            print(f"ERROR: {item} は重複です")
            return False
    
    def remove(self, item):
        # 指定された要素をセットから削除する。
        if item in self._data:
            self._data.remove(item)
            return True
        else:
            print(f"ERROR: {item} は範囲外です")
            return False
    
    def is_empty(self):
        # 空かどうかをチェックする
        return len(self._data) == 0
    
    def size(self):
        # 要素数を返す
        return len(self._data)
    
    def clear(self):
        # 空にする
        self._data.clear()
        return True

def main():
    print("Set TEST -----> start")

    print("\nnew")
    set_data = SetData()
    print(f"  現在のデータ: {set_data.get()}")

    print("\nadd")
    input = [10, 20, 30, 20, 40]
    for item in input:
        print(f"  入力値: {item}")
        output = set_data.add(item)
        print(f"  出力値: {output}")
        print(f"  現在のデータ: {set_data.get()}")

    print("\nsize")
    output = set_data.size()
    print(f"  出力値: {output}")

    print("\nis_empty")
    output = set_data.is_empty()
    print(f"  出力値: {output}")

    print("\nget_value")
    input = [0, 2, 5]
    for index in input:
        print(f"  入力値: {index}")
        output = set_data.get_value(index)
        print(f"  出力値: {output}")

    print("\nget_index")
    input = [30, 99]
    for item in input:
        print(f"  入力値: {item}")
        output = set_data.get_index(item)
        print(f"  出力値: {output}")

    print("\nremove")
    input = [20, 50, 10]
    for item in input:
        print(f"  入力値: {item}")
        output = set_data.remove(item)
        print(f"  出力値: {output}")
        print(f"  現在のデータ: {set_data.get()}")
    
    print("\nsize")
    output = set_data.size()
    print(f"  出力値: {output}")

    print("\nclear")
    output = set_data.clear()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {set_data.get()}")

    print("\nis_empty")
    output = set_data.is_empty()
    print(f"  出力値: {output}")

    print("\nSet TEST <----- end")

if __name__ == "__main__":
    main()

File: data_structure/set/src/SetDemo.swift
// Swift
// データ構造: セット (Set)

import Foundation

class SetData<T: Hashable> {
    private var data: Set<T>
    
    init() {
        data = []
    }
    
    func get() -> [T] {
        return Array(data)
    }
    
    func getIndex(_ item: T) -> Int? {
        guard let index = Array(data).firstIndex(of: item) else {
            print("ERROR: \(item) は範囲外です")
            return nil
        }
        return index
    }
    
    func getValue(at index: Int) -> T? {
        let dataArray = Array(data)
        guard index >= 0 && index < dataArray.count else {
            print("ERROR: \(index) は範囲外です")
            return nil
        }
        return dataArray[index]
    }
    
    func add(_ item: T) -> Bool {
        if !data.contains(item) {
            data.insert(item)
            return true
        } else {
            print("ERROR: \(item) は重複です")
            return false
        }
    }
    
    func remove(_ item: T) -> Bool {
        if data.contains(item) {
            data.remove(item)
            return true
        } else {
            print("ERROR: \(item) は範囲外です")
            return false
        }
    }
    
    func isEmpty() -> Bool {
        return data.isEmpty
    }
    
    func size() -> Int {
        return data.count
    }
    
    func clear() -> Bool {
        data.removeAll()
        return true
    }
}

func main() {
    print("Set TEST -----> start")
    
    print("\nnew")
    let setData = SetData<Int>()
    print("  現在のデータ: \(setData.get())")
    
    print("\nadd")
    let input = [10, 20, 30, 20, 40]
    for item in input {
        print("  入力値: \(item)")
        let output = setData.add(item)
        print("  出力値: \(output)")
        print("  現在のデータ: \(setData.get())")
    }
    
    print("\nsize")
    let sizeOutput = setData.size()
    print("  出力値: \(sizeOutput)")
    
    print("\nis_empty")
    let isEmptyOutput = setData.isEmpty()
    print("  出力値: \(isEmptyOutput)")
    
    print("\nget_value")
    let getValueInput = [0, 2, 5]
    for index in getValueInput {
        print("  入力値: \(index)")
        let output = setData.getValue(at: index)
        print("  出力値: \(output ?? "nil")")
    }
    
    print("\nget_index")
    let getIndexInput = [30, 99]
    for item in getIndexInput {
        print("  入力値: \(item)")
        let output = setData.getIndex(item)
        print("  出力値: \(output ?? -1)")
    }
    
    print("\nremove")
    let removeInput = [20, 50, 10]
    for item in removeInput {
        print("  入力値: \(item)")
        let output = setData.remove(item)
        print("  出力値: \(output)")
        print("  現在のデータ: \(setData.get())")
    }
    
    print("\nsize")
    let finalSizeOutput = setData.size()
    print("  出力値: \(finalSizeOutput)")
    
    print("\nclear")
    let clearOutput = setData.clear()
    print("  出力値: \(clearOutput)")
    print("  現在のデータ: \(setData.get())")
    
    print("\nis_empty")
    let finalIsEmptyOutput = setData.isEmpty()
    print("  出力値: \(finalIsEmptyOutput)")
    
    print("\nSet TEST <----- end")
}

main()